include/gsl/span

```diff
@@ -121,139 +122,165 @@ namespace details
+            if (n > 0) Expects(end_ - current_ >= n);
+            if (n < 0) Expects(end_ - current_ >= -n);
+            current_ += n;
+            return *this;
+        }
+
+        constexpr span_iterator operator+(const difference_type n) const noexcept
+        {
+            auto ret{*this};
             return ret += n;
         }
 
-        friend constexpr span_iterator operator+(difference_type n, span_iterator const& rhs) noexcept
+        friend constexpr span_iterator operator+(const difference_type n,
+                                                 span_iterator const& rhs) noexcept
         {
             return rhs + n;
         }
 
-        constexpr span_iterator& operator+=(difference_type n) noexcept
+        constexpr span_iterator& operator-=(const difference_type n) noexcept
         {
-            Expects((index_ + n) >= 0 && (index_ + n) <= span_->size());
-            index_ += n;
+            Expects(begin_ && end_ && current_);
+            if (n > 0) Expects(end_ - current_ >= n);
+            if (n < 0) Expects(end_ - current_ >= -n);
+            current_ -= n;
             return *this;
         }
 
-        constexpr span_iterator operator-(difference_type n) const noexcept
+        constexpr span_iterator operator-(const difference_type n) const noexcept
         {
-            auto ret = *this;
+            auto ret{*this};
             return ret -= n;
         }
 
-        constexpr span_iterator& operator-=(difference_type n) noexcept { return *this += -n; }
+        friend constexpr span_iterator operator-(const difference_type n,
+                                                 span_iterator const& rhs) noexcept
+        {
+            return rhs - n;
+        }
 
-        constexpr difference_type operator-(span_iterator rhs) const noexcept
+        template <
+            class type2,
+            std::enable_if_t<std::is_same<std::remove_cv_t<type2>, value_type>::value, int> = 0>
+        constexpr difference_type operator-(const span_iterator<type2>& rhs) const noexcept
         {
-            Expects(span_ == rhs.span_);
-            return index_ - rhs.index_;
+            Expects(begin_ == rhs.begin_);
+            return current_ - rhs.current_;
         }
 
-        constexpr reference operator[](difference_type n) const noexcept { return *(*this + n); }
+        constexpr reference operator[](const difference_type n) const noexcept
+        {
+            return *(*this + n);
+        }
 
-        constexpr friend bool operator==(span_iterator lhs, span_iterator rhs) noexcept
+        template <
+            class type2,
+            std::enable_if_t<std::is_same<std::remove_cv_t<type2>, value_type>::value, int> = 0>
+        constexpr bool operator==(const span_iterator<type2>& rhs) const noexcept
         {
-            return lhs.span_ == rhs.span_ && lhs.index_ == rhs.index_;
+            return begin_ == rhs.begin_ && current_ == rhs.current_;
```

Comment:
`&& end_ == rhs.end_` would be consistent here. Given that the standard only requires `x == y` to be valid when `x` and `y` are iterators into the same range, I would write this as `Expects(begin_ == rhs.begin_ && end_ == rhs.end_); return current_ == rhs.current_;`.